compiler/noirc_frontend/src/elaborator/types.rs

```diff
@@ -216,8 +271,13 @@ impl<'context> Elaborator<'context> {
         }
     }
 
-    fn resolve_trait_as_type(&mut self, path: Path, args: Vec<UnresolvedType>) -> Type {
-        let args = vecmap(args, |arg| self.resolve_type_inner(arg));
+    fn resolve_trait_as_type(
+        &mut self,
+        path: Path,
+        args: Vec<UnresolvedType>,
+        kind: TypeKind,
+    ) -> Type {
+        let args = vecmap(args, |arg| self.resolve_type_inner(arg, kind));
```

Comment:
We shouldn't resolve all of the trait args as if they had the same kind when it could change depending on the arg.

Reply:
Switched to now fetch the resolved generics from the the trait and check against the specified kind.